proposals/0018-mqtt-nio.md

```diff
@@ -0,0 +1,136 @@
+```swift
+try await client.publish(
+    to: "TopicName",
+    payload: ByteBuffer(string: "Message Payload"),
+    qos: .atLeastOnce
+)
+```
+Publish messages can be sent with three different levels of QoS: `.atMostOnce`, `.atLeastOnce` and `.exactlyOnce`. The `publish` function will deal with all three QoS levels and only return when all relevant ACK messages have been dealt with.
+
+### Subscribe and Listen
+
+You can send a `SUBSCRIBE` message to the server and listen for `PUBLISH` messages from the server. A listener will respond to all `PUBLISH` messages sent back from the server. If you have multiple subscriptions you will need to check the topic name to verify you have the correct message.
+
+```swift
+let subscription = MQTTSubscribeInfo(topicFilter: "MyTopic", qos: .exactlyOnce)
+// subscribe returns a SUBACK packet which contains the QoS to be expected for each
+// topic subscription
+let suback = try await client.subscribe(to: [subscription])
+// By adding a PUBLISH listener you can process all the PUBLISH messages being sent from
+// the server to the client
+client.addPublishListener("My Listener") { result in
+    switch result {
+    case .success(let publish):
+        if publish.topicName == "MyTopic" {
+            var buffer = publish.payload
+            let string = buffer.readString(length: buffer.readableBytes)
+            print(string)
+        }
+    case .failure(let error):
+        print("Error while receiving PUBLISH event")
+    }
+}
+```
+
+### Version 5
+
+`MQTTNIO` includes support for version 5 features of the MQTT protocol. To connect using the v5 protocol you need to specify this in your `MQTTClient` configuration.
+
+```swift
+let client = MQTTClient(
+    host: "test.mosquitto.org", 
+    port: 1883,
+    identifier: "My unique identifier",
+    eventLoopGroupProvider: .createNew,
+    configuration: .init(version: .v5_0)
+)
+```
+You can use the standard `MQTTClient` APIs with a v5 client but to access v5 features, including MQTT properties and v5 return codes use `MQTTClient.v5`. This includes versions of the `connect`, `publish`, `subscribe`, `unsubscribe` and `disconnect` functions. Below you will see a `publish` call adding the `contentType` property.
+
+```swift
+let puback = try await client.v5.publish(
+    to: "JSONTest", 
+    payload: payload, 
+    qos: .atLeastOnce, 
+    properties: [.contentType("application/json")]
+)
+```
+Whoever subscribes to the "JSONTest" topic with a v5.0 client will also receive the `contentType` property along with the payload.
+
+## Maturity Justification
+
+Sandbox. The package has one author. It is fairly new, v1.0.0 was released a year ago.  
+
+## Alternatives considered
+
+There are two other major Swift MQTT client implementations [CocoaMQTT](https://github.com/emqx/CocoaMQTT) and  [SwiftMQTT](https://github.com/aciidb0mb3r/SwiftMQTT). Neither of these are built on top of Swift NIO or support v5 of the protocol.
```

Comment:
Looks like CocaoMQTT now supports v5 of the protocol. However it's still not built on NIO or aimed at the server ecosystem, and the SSWG has an explicit aim to allow multiple packages for the same thing

Reply:
Updated